File: python/2022/14_regolith_reservoir/aoc202214.py
"""AoC 14, 2022: Regolith Reservoir."""

# Standard library imports
import pathlib
import sys
from itertools import product

ROCK, SAND, FLOOR = range(3)


def parse_data(puzzle_input):
    """Parse input."""
    return {
        pos: ROCK for line in puzzle_input.split("\n") for pos in parse_structure(line)
    }


def parse_structure(line):
    """Parse one rock structure."""
    corners = line.split(" -> ")
    return [
        pos
        for first, last in zip(corners, corners[1:])
        for pos in parse_line(first, last)
    ]


def parse_line(first, last):
    """Parse one line of rocks."""
    first_col, first_row = [int(coord) for coord in first.split(",")]
    last_col, last_row = [int(coord) for coord in last.split(",")]

    if last_col > first_col or last_row > first_row:
        return product(range(first_col, last_col + 1), range(first_row, last_row + 1))
    elif last_col < first_col:
        return product(range(last_col, first_col + 1), range(first_row, last_row + 1))
    elif last_row < first_row:
        return product(range(first_col, last_col + 1), range(last_row, first_row + 1))


def part1(rocks):
    """Solve part 1."""
    abyss = max(row for _, row in rocks.keys())
    cave = simulate_sand(rocks, source=(500, 0), abyss=abyss)
    return sum(kind == SAND for kind in cave.values())


def part2(rocks):
    """Solve part 2."""
    floor = max(row for _, row in rocks.keys()) + 2
    cave = simulate_sand(
        rocks | {(col, floor): FLOOR for col in range(1000)},
        source=(500, 0),
        abyss=floor,
    )
    return sum(kind == SAND for kind in cave.values())


def simulate_sand(cave, source, abyss):
    """Simulate sand falling from the source."""
    while True:
        sand_pos = pour_sand_unit(cave, source, abyss)
        if sand_pos is None:
            return cave
        cave = cave | {sand_pos: SAND}
        if sand_pos == source:
            return cave


def pour_sand_unit(cave, source, abyss):
    """Pour one unit of sand until it hits something or falls into the abyss."""
    col, source_row = source
    for row in range(source_row, abyss + 1):
        if (col, row + 1) not in cave:
            continue
        elif (col - 1, row + 1) not in cave:
            col -= 1
            continue
        elif (col + 1, row + 1) not in cave:
            col += 1
            continue
        else:
            return (col, row)


def solve(puzzle_input):
    """Solve the puzzle for the given input."""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().rstrip())
        print("\n".join(str(solution) for solution in solutions))

File: elixir/lib/2022/14_regolith_reservoir/aoc202214.ex
defmodule AOC2022.Day14 do
  @moduledoc """
  Advent of Code 2022, day 14: Regolith Reservoir.
  """
  require AOC

  @doc """
  Parse input.
  """
  def parse(puzzle_input) do
    puzzle_input
    |> String.split("\n", trim: true)
    |> Enum.flat_map(&parse_structure/1)
    |> Map.from_keys(:rock)
  end

  @doc """
  Parse one rock structure.

  ## Example:

      iex> parse_structure("4,4 -> 6,4 -> 6,2")
      [{4, 4}, {5, 4}, {6, 4}, {6, 4}, {6, 3}, {6, 2}]
  """
  def parse_structure(line),
    do:
      line
      |> String.split(" -> ", trim: true)
      |> Enum.chunk_every(2, 1, :discard)
      |> Enum.flat_map(fn [first, last] -> parse_line(first, last) end)

  @doc """
  Parse one line of rocks.

  ## Examples:

      iex> parse_line("12,5", "9,5")
      [{12, 5}, {11, 5}, {10, 5}, {9, 5}]
      iex> parse_line("7,3", "7,5")
      [{7, 3}, {7, 4}, {7, 5}]
  """
  def parse_line(first, last) do
    [first_col, first_row] = first |> String.split(",") |> Enum.map(&String.to_integer/1)
    [last_col, last_row] = last |> String.split(",") |> Enum.map(&String.to_integer/1)

    case {last_col - first_col, last_row - first_row} do
      {dcol, drow} when dcol > 0 or drow > 0 ->
        line(first_col..last_col, first_row..last_row)

      {dcol, _} when dcol < 0 ->
        line(first_col..last_col//-1, first_row..last_row)

      {_, drow} when drow < 0 ->
        line(first_col..last_col, first_row..last_row//-1)
    end
  end

  @doc """
  Expand line ranges into points on the line.

  ## Examples:

      iex> line(1..3, 2..2)
      [{1, 2}, {2, 2}, {3, 2}]
      iex> line(4..4, 5..1//-1)
      [{4, 5}, {4, 4}, {4, 3}, {4, 2}, {4, 1}]
  """
  def line(cols, rows), do: for(col <- cols, row <- rows, do: {col, row})

  @doc """
  Solve part 1.
  """
  def part1(rocks) do
    abyss = rocks |> Map.keys() |> Enum.map(&elem(&1, 1)) |> Enum.max()

    simulate_sand(rocks, {500, 0}, abyss) |> Enum.count(fn {_, elem} -> elem == :sand end)
  end

  @doc """
  Solve part 2.
  """
  def part2(rocks) do
    floor = rocks |> Map.keys() |> Enum.map(&elem(&1, 1)) |> Enum.max() |> then(&(&1 + 2))

    0..1000
    |> Enum.reduce(rocks, fn col, cave -> Map.put(cave, {col, floor}, :floor) end)
    |> simulate_sand({500, 0}, floor)
    |> Enum.count(fn {_, elem} -> elem == :sand end)
  end

  @doc """
  Simulate sand

  ## Examples:

      iex> rocks = %{{9, 3} => :rock, {10, 3} => :rock, {11, 3} => :rock}
      iex> simulate_sand(rocks, {10, 0}, 3)
      %{{9, 3} => :rock, {10, 3} => :rock, {11, 3} => :rock, {10, 2} => :sand}
      iex> simulate_sand(rocks, {10, 2}, 3)
      %{{9, 3} => :rock, {10, 3} => :rock, {11, 3} => :rock, {10, 2} => :sand}
      iex> simulate_sand(rocks, {9, 0}, 3)
      %{{9, 3} => :rock, {10, 3} => :rock, {11, 3} => :rock}
  """
  def simulate_sand(rocks, source, abyss) do
    Stream.cycle([source])
    |> Enum.reduce_while(rocks, fn source, cave ->
      case pour_sand_unit(cave, source, abyss) do
        nil -> {:halt, cave}
        ^source -> {:halt, Map.put(cave, source, :sand)}
        sand -> {:cont, Map.put(cave, sand, :sand)}
      end
    end)
  end

  @doc """
  Pour one unit of sand from the source.

  Report nil if the sand falls into the abyss, otherwise the end position.

  ## Examples:

      iex> pour_sand_unit(%{{9, 3} => :rock}, {10, 0}, 3)
      nil
      iex> rocks = %{{9, 3} => :rock, {10, 3} => :rock, {11, 3} => :rock}
      iex> pour_sand_unit(rocks, {10, 0}, 3)
      {10, 2}
      iex> pour_sand_unit(rocks |> Map.put({10, 2}, :sand), {10, 0}, 3)
      nil
  """
  def pour_sand_unit(_, {_, row}, abyss) when row >= abyss, do: nil

  def pour_sand_unit(cave, {col, row}, abyss) do
    cond do
      not Map.has_key?(cave, {col, row + 1}) ->
        pour_sand_unit(cave, {col, row + 1}, abyss)

      not Map.has_key?(cave, {col - 1, row + 1}) ->
        pour_sand_unit(cave, {col - 1, row + 1}, abyss)

      not Map.has_key?(cave, {col + 1, row + 1}) ->
        pour_sand_unit(cave, {col + 1, row + 1}, abyss)

      true ->
        {col, row}
    end
  end

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end
